Dockerfile

```
@@ -0,0 +1,42 @@
+################################################################################################################ builder
+
+FROM maven:3-jdk-8 as builder
+WORKDIR /build
+
+# this block will make docker cache the dependencies for subsequent builds of the same image.
+COPY pom.xml .
+COPY subsonic-rest-api/pom.xml ./subsonic-rest-api/pom.xml
+COPY integration-test/pom.xml ./integration-test/pom.xml
+COPY airsonic-sonos-api/pom.xml ./airsonic-sonos-api/pom.xml
+COPY airsonic-main/pom.xml ./airsonic-main/pom.xml
+RUN mvn de.qaware.maven:go-offline-maven-plugin:resolve-dependencies
+
+COPY . .
+RUN mvn package
+
+################################################################################################################ runtime
+
+FROM openjdk:8-jre-alpine as runtime
+LABEL description="Airsonic is a free, web-based media streamer, providing ubiquitious access to your music." \
+      url="https://github.com/airsonic/airsonic"
+ENV AIRSONIC_PORT=4040 AIRSONIC_DIR=/airsonic CONTEXT_PATH=/ UPNP_PORT=4041 JVM_HEAP=256m
+
+WORKDIR $AIRSONIC_DIR
+RUN apk --no-cache add \
+    ffmpeg \
+    lame \
+    bash \
+    libressl \
+    fontconfig \
+    ttf-dejavu \
+    ca-certificates \
+    tini
+COPY install/docker/run.sh /usr/local/bin/run.sh
+RUN chmod +x /usr/local/bin/run.sh
+COPY --from=builder  /build/airsonic-main/target/airsonic.war airsonic.war
+EXPOSE $AIRSONIC_PORT
+EXPOSE $UPNP_PORT
+EXPOSE 1900/udp
```

> The EXPOSE instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published.

https://docs.docker.com/engine/reference/builder/#expose

I'd also suggest that we get rid of what's on L22 for ports and just expose what we know is certain here.

```suggestion
EXPOSE 4040
EXPOSE 4041
EXPOSE 1900/udp
```

---

Users will be able to expose what they want on their host, but keeping this as default inside the container won't interfere with other containers or the user's port mappings.

Example:

```shell script
docker run -it --rm \
    -p $MY_AIRSONIC_PORT:4040 \
    -e $MY_AIRSONIC_CONTEXT_PATH:/
    airsonic
```

---

I disagree about removing the ENVs. They're not there solely for the EXPOSE portion of the Dockerfile (which as noted is purely for optics). They're actually there to provide the artifact running inside the container some default and needed env variables.

In the example command provided above, sure the container published 4040  to host's $MY_AIRSONIC_PORT, but the problem is that the artifact inside the container doesn't know that it needs to bind to the container's 4040 (same with the UPNP port etc). It could be binding to the container's 8080, in which case the wrong port is being published by the container to host.

---

Oh -- I didn't know they were used in code. That makes sense. Linking to what you've described.

https://github.com/airsonic/airsonic/blob/bd83229345976bd92446a02c30656adbb0de13a9/install/docker/run.sh#L19-L22

Thanks!

---

Looks like this one is resolved.